sklearn/metrics/pairwise.py

```diff
@@ -224,41 +234,156 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     """
     X, Y = check_pairwise_arrays(X, Y)
 
-    if X_norm_squared is not None:
-        XX = check_array(X_norm_squared)
-        if XX.shape == (1, X.shape[0]):
-            XX = XX.T
-        elif XX.shape != (X.shape[0], 1):
-            raise ValueError(
-                "Incompatible dimensions for X and X_norm_squared")
-    else:
-        XX = row_norms(X, squared=True)[:, np.newaxis]
+    XX, YY = _check_norms(X, Y, X_norm_squared, Y_norm_squared)
+
+    n_features = X.shape[1]
+
+    # For n_features > 32 we use the 'fast 'method to compute the euclidean
+    # distance, i.e. d(x,y)² = ||x||² + ||y||² - 2 * x.y
+    # It's faster but less precise.
+    if n_features > 32:
 
-    if X is Y:  # shortcut in the common case euclidean_distances(X, X)
-        YY = XX.T
-    elif Y_norm_squared is not None:
-        YY = np.atleast_2d(Y_norm_squared)
+        # To minimize precision issues with float32, we compute the distance
+        # matrix on chunks of X and Y upcasted to float64
+        if X.dtype == np.float32:
```

Comment:
If I may add a few comments on the code inspired by mine.

What happens in case of mixed type between `X` and `Y`? I guess the expected result is a float64 matrix. Testing only the type of `X` seems fishy.

Reply:
See `_return_float_dtype` above used in `check_pairwise_arrays`, called here.

Reply:
You can't have mixed dtypes of X and Y due to `check_pairwise_arrays` as told above, this is why your code could be a little bit simplified.

Reply:
Indeed. My bad.